### PR TITLE
feat(librechat-code-interpreter): self-host LibreChat Code Interpreter (ADR-0122)

### PR DESCRIPTION
**Pull request by stephane-segning**
## 1. Summary

This PR changes:

- Adds `charts/librechat-code-interpreter` — the self-hosted [`clickhouse/code-interpreter`](https://github.com/clickhouse/code-interpreter) service (5 components: API, service-worker + sandbox-runner, file-server, tool-call-server, egress-gateway), vendored from upstream's `helm/codeapi` and adapted to repo convention (no bundled Redis/MinIO, ExternalSecret instead of a plaintext Secret template, NsJail sandbox mode).
- Deploys it as a flat `charts/apps` app in a new dedicated `librechat-sandbox` namespace, elevated to the `privileged` Pod Security Standard (scoped away from `converse`).
- `charts/librechat-app`: `LIBRECHAT_CODE_BASEURL` now points at the in-cluster service; adds `CODEAPI_JWT_*` env + a new ExternalSecret so LibreChat mints short-lived Ed25519-signed JWTs per request (the self-hosted service rejects a static API key outside local mode).
- ADR-0122 + `docs/patterns/self-hosted-code-interpreter.md` (the secret-generation runbook) + index updates.

It solves:

- We previously wired LibreChat's Code Interpreter capability to the *managed* librechat.ai sandbox API (#734) — code ran on their infra, not ours. This replaces that with our own instance.

---

## 2. Intent

> Run agent code execution on our own cluster instead of depending on a third-party managed API, and set up the prerequisites for future Programmatic Tool Calling (MCP tools called from sandboxed code).

---

### In Scope

- The `librechat-code-interpreter` chart itself (vendored + adapted).
- Wiring it into `charts/apps` (namespace, PSS, `depsOverlay`).
- The `librechat-app` side of JWT auth (env + ExternalSecret).
- ADR-0122 + the operational runbook doc + doc-index updates.

### Out of Scope

- The matching Cilium `CiliumNetworkPolicy` overlay (file-server → Hetzner Object Storage) — drafted locally but lives in the **separate, private `ai-helm-values` repo** (`environments/{base,prod}/deps/librechat-code-interpreter/`), not committed there yet.
- Generating/storing the actual secret material in `ssegning-aws` — out of band, see the runbook doc.
- KVM microVM sandbox mode (needs confirmed `/dev/kvm` on a `home-remote` node pool — noted as a follow-up).
- Hardening `api`/`file-server`/`tool-call-server`/`egress-gateway` container `securityContext` to this repo's usual KSV-0118 profile — deferred until each image's non-root behaviour is confirmed live.

---

## 4. Verification

I verified this change by:

- [x] Running automated tests (`helm lint --strict` + `helm template --dry-run`)
- [ ] Running manual tests
- [ ] Checking logs
- [ ] Checking metrics
- [ ] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```bash
helm dep build charts/librechat-code-interpreter charts/apps charts/librechat-app charts/librechart
helm lint charts/librechat-code-interpreter charts/apps charts/librechat-app charts/librechart --strict
helm template codeapi charts/librechat-code-interpreter -n librechat-sandbox --dry-run
helm template apps charts/apps --dry-run
helm template lc charts/librechat-app -n converse --dry-run
```

Results:

```text
0 chart(s) failed on all four charts (librechat-code-interpreter, apps, librechat-app, librechart).
Rendered manifests confirmed: 6 Deployments/3 Services/6 NetworkPolicies/1 PVC/1 Job/1 ExternalSecret
in librechat-code-interpreter; the new aii-librechat-code-interpreter Application with the
librechat-sandbox privileged PSS label in apps; CODEAPI_JWT_* env + librechat-codeapi-jwt
ExternalSecret scaffold in librechat-app.
```

No cluster access in this session, so nothing was verified live — see the "Out of Scope" and the runbook's "Verifying a fresh install" section for what a reviewer with cluster access should check post-merge.

---

## 5. Screenshots / Evidence

* Logs: `helm lint`/`helm template` output pasted above.

---

## 6. Risk Assessment

Risk level:

* [ ] Low
* [x] Medium
* [ ] High

Potential risks:

* NsJail/chroot sandbox mode grants the sandbox-runner container `SYS_ADMIN`/`NET_ADMIN`/`SYS_PTRACE`/`seccompProfile: Unconfined` + 8 more capabilities, in a namespace elevated to the `privileged` Pod Security Standard — meaningfully more privilege than the KVM microVM path would need. This is a deliberate trade-off (no confirmed `/dev/kvm` on the worker pool), scoped to a brand-new namespace that runs nothing else.
* First install won't work at all until the 8 `ssegning-aws` properties exist (see the runbook) and the `ai-helm-values` Cilium overlay is committed there — both out-of-band follow-ups, not part of this PR.
* `sandboxRunner`/`api` pinned to 1 replica (no HA) due to RWO-only storage on the general worker pool.

Mitigation:

* Sandbox privilege elevation is isolated to its own namespace (`librechat-sandbox`), not the shared `converse` namespace.
* ADR-0122 documents the trade-off and the follow-up path (flip to KVM mode once `/dev/kvm` is confirmed on a node pool).
* Runbook doc gives exact verification commands for a post-merge live check.

---

## 7. AI Usage Declaration

AI was used for:

* [x] Understanding existing code
* [x] Generating code
* [x] Drafting documentation
* [ ] Refactoring
* [ ] Generating tests
* [ ] Reviewing the diff
* [ ] Not used

Human verification:

* [ ] I understand every meaningful change in this PR
* [ ] I checked generated code manually
* [ ] I checked generated tests manually
* [ ] I removed unsupported AI assumptions
* [ ] I accept responsibility for this PR

*(This PR was authored end-to-end by Claude Code in an assistant session; the human maintainer should complete the verification checklist above before merging.)*

---

## 8. Reviewer Focus

Please focus your review on:

* [x] Correctness
* [x] Architecture
* [x] Security
* [ ] Performance
* [ ] Tests
* [x] Maintainability
* [ ] Product intent
* [x] Edge cases

Specifically: the NsJail capability grant + namespace isolation, the JWT auth wiring on the `librechat-app` side (keys/kid must match `charts/librechat-code-interpreter`'s `api.extraEnv`), and whether reusing the shared S3 bucket/redis-ha is acceptable vs. dedicated infra.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: charts/librechat-code-interpreter/values.yaml [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+# =============================================================================
+# REDIS — always external (repo convention: reuse the shared redis-ha,
+# never bundle a Redis chart). See CLAUDE.md "Redis — deployed by the
+# home-os repo". TLS-only, like every other redis-ha consumer.
+# =============================================================================
+redis:
+  enabled: false
+  external:
+    host: redis-ha-redis-haproxy.redis-system.svc.cluster.local
+    port: 6379
```
[the remark this answers is not in the document]
Comment (stephane-segning) [reply charts/librechat-code-interpreter/values.yaml]: Confirmed and fixed in f191fe1 — every component (api, service-worker, file-server, tool-call-server, egress-gateway) now sets REDIS_TLS=true via a new extraEnv hook (three of the five templates had none at all, added them). Note this service only supports encrypted-but-unauthenticated TLS (tls.rejectUnauthorized:false in its own source) — it has no CA-verification knob like librechat-app's rediss://+REDIS_CA, so unlike that path this cannot be upgraded to full CA verification without patching upstream. Documented as an accepted gap (in-cluster only) in values.yaml.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: charts/librechat-code-interpreter/values.yaml [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+            REDIS_PASSWORD:
+              secretKeyRef: { name: codeapi-secrets, key: redis-password }
+            # redis-ha is TLS-only (port 0 / tls-port 6379) — a plaintext
+            # client gets "connection reset by peer". This service's own
+            # Redis TLS support (service/src/queue.ts et al.) is
+            # REDIS_TLS=true -> tls.rejectUnauthorized:false — encrypted but
+            # NOT CA-verified (no REDIS_CA/NODE_EXTRA_CA_CERTS equivalent
+            # upstream, unlike librechat-app's rediss://+REDIS_CA). Accepted
+            # for now: redis-ha is in-cluster only, no public exposure.
+            REDIS_TLS: "true"
```
[the remark this answers is not in the document]
Comment (stephane-segning) [reply charts/librechat-code-interpreter/values.yaml:131]: Agreed, tracked rather than fixed here — added an explicit Neutral/follow-ups entry in ADR-0122 spelling out the risk (Redis carries the job queue, internal service token, execution manifests, egress-grant ledger) and why it's accepted for now (needs an upstream patch for a CA-verification knob; sandbox-runner's own egress is locked to egress-gateway only, so untrusted code can't reach Redis directly). Not changed in code — this service genuinely has no rejectUnauthorized:true/CA-pinning option today.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .trivyignore [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# codeapi.controllers.sandbox-runner.containers.sandbox-runner.securityContext
+# for the full capability list), which forces a non-default securityContext
+# (KSV-0118) and rules out readOnlyRootFilesystem (KSV-0014 — NsJail writes
+# its per-job workspace/scratch dirs to the container rootfs). Every OTHER
+# controller in this chart (api, file-server, tool-call-server,
+# egress-gateway, service-worker, both busybox init containers, package-init)
+# carries this repo's usual hardened securityContext (drop ALL, no priv-esc,
+# readOnlyRootFilesystem where the container can tolerate it) — see the same
+# values.yaml, not ignored.
+KSV-0005
```
[the remark this answers is not in the document]
Comment (stephane-segning) [reply .trivyignore]: Confirmed and fixed — dropped all three KSV-0005/0014/0118 lines from the plain .trivyignore. You're right: once the chart gained the bjw-template + common dependency, security-gates.yml's bare-checkout render can never reach it, so those entries were dead weight carrying only the blind-spot risk you described. The chart stays fully covered via the scoped .trivyignore.yaml entries (release-helm-charts.yml, which does build deps). Kept the file itself with its explanatory header for next time this actually applies to a dependency-free chart.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
